File: single-source-shortest-paths/golang/bellman_ford.go
package main

import "fmt"

func bellmanFord(g *Graph, s *Vertex) bool {
	initializeSingleSource(g, s)

	vertices := g.vertices
	edges := g.edges
	for i := range len(vertices) - 1 {
		fmt.Printf("--- i = %d ---\n", i)

		for _, edge := range edges {
			start := edge.start
			end := edge.end
			weight := edge.weight

			relax(vertices[start], vertices[end], weight)
		}
	}

	hasCycle := false
	for _, edge := range edges {
		start := edge.start
		end := edge.end
		weight := edge.weight

		if vertices[end].distance > vertices[start].distance+weight {
			hasCycle = true

			break
		}
	}

	return hasCycle
}

func printReseult(s *Vertex, v *Vertex, v_name string) {
	if v == s {
		fmt.Printf("%s", s.name)
		if v.name != v_name {
			fmt.Print(" -> ")
		}
	} else if v.predecessor != nil {
		printReseult(s, v.predecessor, v_name)
		fmt.Printf("%s", v.name)
		if v.name != v_name {
			fmt.Print(" -> ")
		}
	}
}

func main() {
	vertices := make(map[string]*Vertex)
	for _, name := range "styxz" {
		vertices[string(name)] = &Vertex{name: string(name)}
	}

	edges := make([]*Edge, 0)
	edges = append(edges, &Edge{start: "s", end: "t", weight: 6})
	edges = append(edges, &Edge{start: "s", end: "y", weight: 7})
	edges = append(edges, &Edge{start: "t", end: "y", weight: 8})
	edges = append(edges, &Edge{start: "t", end: "x", weight: 5})
	edges = append(edges, &Edge{start: "t", end: "z", weight: -4})
	edges = append(edges, &Edge{start: "y", end: "x", weight: -3})
	edges = append(edges, &Edge{start: "y", end: "z", weight: 9})
	edges = append(edges, &Edge{start: "x", end: "t", weight: -2})
	edges = append(edges, &Edge{start: "z", end: "s", weight: 2})
	edges = append(edges, &Edge{start: "z", end: "x", weight: 7})

	graph := &Graph{vertices: vertices, edges: edges}
	graph.initializeGraph()

	fmt.Println("--- graph ---")
	graph.printGraph()

	fmt.Println("--- Bellman-Ford algorithm ---")
	hasCycle := bellmanFord(graph, vertices["s"])

	fmt.Println("--- result ---")
	if hasCycle {
		fmt.Println("Graph has a cycle.")
	} else {
		for _, v := range vertices {
			fmt.Printf("%s (%d): ", v.name, v.distance)
			printReseult(vertices["s"], v, v.name)
			fmt.Println()
		}
	}
}

File: single-source-shortest-paths/python/bellman_ford.py
from graph import Edge, Graph, Vertex, initialize_single_source, relax


def bellman_ford(graph, s):
    initialize_single_source(graph, s)

    vertices = graph.vertices
    edges = graph.edges

    for i in range(len(vertices) - 1):
        print(f"--- i = {i} ---")
        for edge in edges:
            start = edge.start
            end = edge.end
            weight = edge.weight
            relax(vertices[start], vertices[end], weight)

    has_cycle = False
    for edge in edges:
        start = edge.start
        end = edge.end
        weight = edge.weight
        if vertices[end].distance > vertices[start].distance + weight:
            has_cycle = True

            break

    return has_cycle


def print_path(s, v, v_name):
    if v == s:
        print(s.name, end=" -> " if v.name != v_name else "")
    elif v.predecessor:
        print_path(s, v.predecessor, v_name)
        print(v.name, end=" -> " if v.name != v_name else "")


if __name__ == "__main__":
    vertices = {}
    for name in "styxz":
        vertices[name] = Vertex(name)

    edges = []
    edges.append(Edge("s", "t", 6))
    edges.append(Edge("s", "y", 7))
    edges.append(Edge("t", "y", 8))
    edges.append(Edge("t", "x", 5))
    edges.append(Edge("t", "z", -4))
    edges.append(Edge("y", "x", -3))
    edges.append(Edge("y", "z", 9))
    edges.append(Edge("x", "t", -2))
    edges.append(Edge("z", "s", 2))
    edges.append(Edge("z", "x", 7))

    graph = Graph(vertices, edges)

    print("--- graph ---")
    graph.print_graph()

    print("--- Bellman-Ford algorithm ---")
    has_cycle = bellman_ford(graph, vertices["s"])

    print("--- result ---")
    if has_cycle:
        print("Graph has a cycle.")
    else:
        for v in vertices.values():
            print(f"{v.name} : {v.distance} ", end="(")
            print_path(vertices["s"], v, v.name)
            print(")")
